File: apps/sdk-python/src/praxis_sdk/services/identity.py
"""``IdentityService`` — typed client for the ``agent-identity`` service.

Mirror of ``apps/sdk-typescript/src/services/identity.ts`` and
``apps/agent-identity/src/http/routes.ts``:

- ``POST /v1/identity/register``
- ``GET  /v1/identity/:did``
- ``POST /v1/identity/verify``
"""

from __future__ import annotations

from dataclasses import dataclass

from .._http import HttpClientOptions, RequestParams, request
from ._convert import from_wire, to_wire


@dataclass(frozen=True, slots=True)
class RegisterResponse:
    """Response of ``POST /v1/identity/register``."""

    did: str
    agent_id: str
    registered_at: str


@dataclass(frozen=True, slots=True)
class ResolveResponse:
    """Response of ``GET /v1/identity/:did``."""

    did: str
    agent_id: str
    public_key: str
    signature_scheme: str
    owner_operator_id: str
    registered_at: str
    revoked_at: str | None = None


@dataclass(frozen=True, slots=True)
class VerifyResponse:
    """Response of ``POST /v1/identity/verify``."""

    valid: bool
    reason: str | None = None


class IdentityService:
    """Typed client for the ``agent-identity`` service."""

    def __init__(self, opts: HttpClientOptions, base_url: str) -> None:
        self._opts = opts
        self._base_url = base_url

    def register(self, *, public_key: str, owner_operator_id: str) -> RegisterResponse:
        """``POST /v1/identity/register``.

        Args:
            public_key: Ed25519 public key, raw 32 bytes, base64-encoded.
            owner_operator_id: Operator that owns this agent (1..128 chars).
        """
        body = to_wire({"public_key": public_key, "owner_operator_id": owner_operator_id})
        data = request(
            self._opts,
            RequestParams(
                method="POST",
                base_url=self._base_url,
                path="/v1/identity/register",
                body=body,
            ),
        )
        if data is None:
            raise RuntimeError("identity.register: empty response body")
        return from_wire(RegisterResponse, data)

    def resolve(self, did: str) -> ResolveResponse:
        """``GET /v1/identity/:did``."""
        from urllib.parse import quote

        data = request(
            self._opts,
            RequestParams(
                method="GET",
                base_url=self._base_url,
                path=f"/v1/identity/{quote(did, safe='')}",
            ),
        )
        if data is None:
            raise RuntimeError("identity.resolve: empty response body")
        return from_wire(ResolveResponse, data)

    def verify(self, *, did: str, message: str, signature: str) -> VerifyResponse:
        """``POST /v1/identity/verify``.

        Args:
            did: The signer's ``did:key:z...`` identifier.
            message: base64-encoded message bytes.
            signature: base64-encoded Ed25519 signature.
        """
        body = to_wire({"did": did, "message": message, "signature": signature})
        data = request(
            self._opts,
            RequestParams(
                method="POST",
                base_url=self._base_url,
                path="/v1/identity/verify",
                body=body,
            ),
        )
        if data is None:
            raise RuntimeError("identity.verify: empty response body")
        return from_wire(VerifyResponse, data)

File: apps/sdk-typescript/src/services/identity.ts
/**
 * `IdentityService` — typed client for the `agent-identity` service.
 *
 * Mirror of `apps/agent-identity/src/http/routes.ts`:
 *   - POST /v1/identity/register
 *   - GET  /v1/identity/:did
 *   - POST /v1/identity/verify
 */

import { request } from '../http.js';

import type { HttpClientOptions } from '../http.js';

export interface RegisterRequest {
  /** Ed25519 public key, raw 32 bytes, base64-encoded. */
  readonly publicKey: string;
  /** Operator that owns this agent (1..128 chars). */
  readonly ownerOperatorId: string;
}

export interface RegisterResponse {
  readonly did: string;
  readonly agentId: string;
  readonly registeredAt: string;
}

export interface ResolveResponse {
  readonly did: string;
  readonly agentId: string;
  readonly publicKey: string;
  readonly signatureScheme: string;
  readonly ownerOperatorId: string;
  readonly registeredAt: string;
  readonly revokedAt: string | null;
}

export interface VerifyRequest {
  readonly did: string;
  /** base64-encoded message bytes. */
  readonly message: string;
  /** base64-encoded signature bytes. */
  readonly signature: string;
}

export interface VerifyResponse {
  readonly valid: boolean;
  readonly reason?: string;
}

export class IdentityService {
  constructor(
    private readonly opts: HttpClientOptions,
    private readonly baseUrl: string,
  ) {}

  /** POST /v1/identity/register */
  public async register(body: RegisterRequest): Promise<RegisterResponse> {
    const data = await request<RegisterResponse>(this.opts, {
      method: 'POST',
      baseUrl: this.baseUrl,
      path: '/v1/identity/register',
      body,
    });
    if (data === undefined) {
      throw new Error('identity.register: empty response body');
    }
    return data;
  }

  /** GET /v1/identity/:did */
  public async resolve(did: string): Promise<ResolveResponse> {
    const data = await request<ResolveResponse>(this.opts, {
      method: 'GET',
      baseUrl: this.baseUrl,
      path: `/v1/identity/${encodeURIComponent(did)}`,
    });
    if (data === undefined) {
      throw new Error('identity.resolve: empty response body');
    }
    return data;
  }

  /** POST /v1/identity/verify */
  public async verify(body: VerifyRequest): Promise<VerifyResponse> {
    const data = await request<VerifyResponse>(this.opts, {
      method: 'POST',
      baseUrl: this.baseUrl,
      path: '/v1/identity/verify',
      body,
    });
    if (data === undefined) {
      throw new Error('identity.verify: empty response body');
    }
    return data;
  }
}
